In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import itertools

%matplotlib inline

In [2]:
mnist = pd.read_csv("data/mnist1.5k.csv.gz", compression="gzip", header=None).values
X=mnist[:,1:785]/255.
y=mnist[:,0]
print("dimension de las imagenes y las clases", X.shape, y.shape)

dimension de las imagenes y las clases (1500, 784) (1500,)


In [3]:
X_train, X_test, y_train, y_test = X[:300], X[300:], y[:300], y[300:]
y_train_oh = np.eye(10)[y_train]
y_test_oh  = np.eye(10)[y_test]
print(X_train.shape, y_train_oh.shape,  X_test.shape, y_test_oh.shape)

(300, 784) (300, 10) (1200, 784) (1200, 10)


## Network architectures lab


from the notes we have two architectures:
    
- **A**: Three layer network accepting a 784 element vector as input and outputing a 10-class vector
- **B**: Same as **A** but accepts an additional 2 element vector with _evenness_ information that is injected at the third layer.

This lab requires you to do two things:

**1**. Create a **C** architecture similar to **B** but where the 2 element vector is injected at the second layer. This will make the network have 41,650 parameters, distributed in the following way:
   
   

        INPUT 1 to LAYER 1:              784*50 + 50 (bias) = 39250
        LAYER 1 to LAYER 2:               50*30 + 30 (bias) = 1530
        LAYER 2 + INPUT 2 to LAYER 3: (30+2)*20 + 20 (bias) = 660
        LAYER 3 to OUTPUT:                20*10 + 10 (bias) = 210
    
                                                       TOTAL 41650
                                                       
**2**. Run an experimental setup where you train different network configurations and measure the accuracy on test data. Fix the number of neurons to 50, 30 and 20 for each layer and the following combination of parameters:

- For architecture **A** (3 configurations)

    `s3_activation` $\in$ `["linear", "relu", "tanh"]`

- For architectures **B** and **C** (15 configurations for each architecture)

    `s3_activation` $\in$ `["linear", "relu", "tanh"]`
    
    `k1,k2` $\in$ `[(0,1), (-.5,2),(-.5,30), (0,15),(0,30)]`

And create a heat map showing the accuracy in test obtained for each configuration, such as the following (your results should be approximate to this):

![alt text](./Images/mm_results_1.png)


And two bar plots with the average per architecture and k1,k2 configuration, such as the following (again, your results should be approximetely similar):

![alt text](./Images/mm_results_2.png)

### Model definitions

In [17]:
from tensorflow.keras import Sequential, Model
from tensorflow.keras.layers import Dense, Dropout, Flatten, concatenate, Input
from tensorflow.keras.backend import clear_session

In [6]:
def get_model_A(input_dim, s1, s2, s3, s3_activation="relu"):
    clear_session()
    model = Sequential()

    ... COPY FROM NOTES ...
    
    return model

In [7]:
def get_model_B(input_dim, extra_info_dim,  s1, s2, s3, s3_activation="relu"):
    clear_session()

    ... COPY FROM NOTES ...    
    
    return model 

In [8]:
def get_model_C(input_dim, extra_info_dim, s1, s2, s3, s3_activation="relu"):
    clear_session()

    
    ... CREATE MODEL ...    
    
    return model

In [9]:
def get_X_extra(y_train, y_test, k0, k1):
    X_train_extra = (np.eye(2)[y_train%2]+k0)*k1
    X_test_extra  = (np.eye(2)[y_test%2]+k0)*k1
    return X_train_extra, X_test_extra

### Use the following dataframe to record your data 

In [3]:
k_set     = [(0,1), (-.5,2),(-.5,30), (0,15),(0,30), None]
act_set   = ["linear", "relu", "tanh"]
arch_set  = ["A", "B", "C"]


r_test = pd.DataFrame(np.zeros((len(arch_set)*len(act_set), len(k_set)))*np.nan, 
                      index=[[a+"-"+b for a,b in itertools.product (arch_set, act_set)]],
                      columns=[str(i) for i in k_set])
r_test

(0, 1)  (-0.5, 2)  (-0.5, 30)  (0, 15)  (0, 30)  None
A-linear     NaN        NaN         NaN      NaN      NaN   NaN
A-relu       NaN        NaN         NaN      NaN      NaN   NaN
A-tanh       NaN        NaN         NaN      NaN      NaN   NaN
B-linear     NaN        NaN         NaN      NaN      NaN   NaN
B-relu       NaN        NaN         NaN      NaN      NaN   NaN
B-tanh       NaN        NaN         NaN      NaN      NaN   NaN
C-linear     NaN        NaN         NaN      NaN      NaN   NaN
C-relu       NaN        NaN         NaN      NaN      NaN   NaN
C-tanh       NaN        NaN         NaN      NaN      NaN   NaN

### Build a loop over the configurations

In [ ]:
for k, act, arch in itertools.product(k_set, act_set, arch_set):
    
    print "%10s %8s %1s"%(str(k), act, arch)    
    
    ... PERFORM EXPERIMENT ACCORDING TO PARAMETERS ...
    



In [ ]:
r_test

### Build the heatmap

In [ ]:
plt.figure(figsize=(8,3))

... YOUR CODE HERE ...

### Build the bar plots

In [ ]:
plt.figure(figsize=(15,4))
plt.subplot(121)

... YOUR CODE HERE ...

plt.subplot(122)

... YOUR CODE HERE ...